linera-execution/src/lib.rs

```diff
@@ -50,6 +51,136 @@ use serde::{Deserialize, Serialize};
+        .map_err(|_| ArithmeticError::Underflow)?)
+}
+
+/// The entries of the runtime related to storage
+#[derive(Copy, Debug, Clone)]
+pub struct RuntimeLimits {
+    /// maximum_budget for reading
+    pub max_budget_num_reads: u64,
+    /// maximum budget of bytes read
+    pub max_budget_bytes_read: u64,
+    /// maximum budget of bytes written
+    pub max_budget_bytes_written: u64,
+    /// The maximum size of read allowed per block
+    pub maximum_bytes_read: u64,
+    /// The maximum size of write allowed per block
+    pub maximum_bytes_written: u64,
+}
+
+/// The entries of the runtime related to storage
+#[derive(Copy, Debug, Clone)]
+pub struct ResourceTracker {
+    /// The used fuel in the computation
+    pub used_fuel: u64,
+    /// The number of reads in the computation
+    pub num_reads: u64,
+    /// The total number of bytes read
+    pub bytes_read: u64,
+    /// The total number of bytes written
+    pub bytes_written: u64,
+    /// The maximum size of read that remains available for use
+    pub maximum_bytes_read: u64,
+    /// The maximum size of write that remains available for use
+    pub maximum_bytes_written: u64,
+}
+
+#[cfg(any(test, feature = "test"))]
+impl Default for ResourceTracker {
+    fn default() -> Self {
+        ResourceTracker {
+            used_fuel: 0,
+            num_reads: 0,
+            bytes_read: 0,
+            bytes_written: 0,
+            maximum_bytes_read: u64::MAX / 2,
+            maximum_bytes_written: u64::MAX / 2,
+        }
+    }
+}
+
+impl Default for RuntimeLimits {
+    fn default() -> Self {
+        RuntimeLimits {
+            max_budget_num_reads: u64::MAX / 2,
+            max_budget_bytes_read: u64::MAX / 2,
+            max_budget_bytes_written: u64::MAX / 2,
+            maximum_bytes_read: u64::MAX / 2,
+            maximum_bytes_written: u64::MAX / 2,
+        }
+    }
+}
+
+impl ResourceTracker {
+    /// Update the limits for the maximum and update the balance.
+    /// The substraction are guaranteed to work correctly and the
+    /// checks are occurring in the runtime.
+    pub fn update_limits(
+        &mut self,
+        balance: &mut Amount,
+        policy: &ResourceControlPolicy,
+        runtime_counts: RuntimeCounts,
+    ) -> Result<(), ExecutionError> {
+        // The fuel being used
+        let initial_fuel = policy.remaining_fuel(*balance);
+        let used_fuel = initial_fuel.saturating_sub(runtime_counts.remaining_fuel);
+        self.used_fuel += used_fuel;
+        sub_assign_fees(balance, policy.fuel_price(used_fuel)?)?;
+        // The number of reads
+        sub_assign_fees(
+            balance,
+            policy.storage_num_reads_price(&runtime_counts.num_reads)?,
+        )?;
+        self.num_reads += runtime_counts.num_reads;
+        // The number of bytes read
+        let bytes_read = runtime_counts.bytes_read;
+        self.maximum_bytes_read -= bytes_read;
+        self.bytes_read += runtime_counts.bytes_read;
```

Comment:
Is it correct to both remove them from the maximum _and_ add them to `bytes_read`?

Reply:
Yes, because the `bytes_read` as a control variable that could be used for `Prometheus`.